drv/gimlet-seq-server/src/vcore.rs

```diff
@@ -0,0 +1,137 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+///
+/// We have seen adventures on the V12_SYS_A2 rail in that it will sag from
+/// 12V to ~8V over a period of about ~4ms, and then rise back 12V over ~7ms.
+/// This happens only on very few machines, and even then happens very rarely
+/// (happening once over hours or days), but the consequences are acute:  the
+/// dip in power results in U.2 drives resetting, and ultimately, the system
+/// resetting itself.  To better characterize any such dips, we want to use
+/// one of the rails on of the RAA229618s (specifically, VDD_VCORE) as a
+/// witness to any V12_SYS_A2 rail fluctuation via its VIN: we set its VIN
+/// undervoltage warning limit to a value that is lower than any we expect in
+/// an operable system (but higher than the sags we have observed), and then
+/// configure its fault output (PWR_CONT1_VCORE_TO_SP_ALERT_L, connected to
+/// PI14) to generate an interrupt on a falling edge.  Upon the interrupt, we
+/// will get notification here, and we will record values of VIN as quickly as
+/// we can.  Each READ_VIN requires 8 bytes, over 3 I2C transactions:
+///
+///   [Write + PAGE + rail] [Write + READ_VIN] [Read + MSB + LSB]
+///
+/// At our midbus speed of 100kHz, this is ~900µs per READ_VIN.  We gather 50
+/// of these READ_VIN measurements, along with timestamps before and after the
+/// operations, and put them all in a ring buffer.  Note that we don't clear
+/// faults after this condition; we will wait until the machine next makes an
+/// A2 to A0 transition to clear faults.
+///
+use drv_i2c_api::{I2cDevice, ResponseCode};
+use drv_i2c_devices::raa229618::Raa229618;
+use drv_stm32xx_sys_api as sys_api;
+use ringbuf::*;
+use sys_api::IrqControl;
+use userlib::*;
+
+pub struct VCore {
+    device: Raa229618,
+    sys: sys_api::Sys,
+}
+
+#[derive(Copy, Clone, PartialEq)]
+enum Trace {
+    Notified,
+    Fault,
+    Start(u64),
+    Reading(units::Volts),
+    Error(ResponseCode),
+    Done(u64),
+    None,
+}
+
+ringbuf!(Trace, 120, Trace::None);
+
+///
+/// We are going to set our input undervoltage warn limit to be 11.75 volts.
+/// Note that we will not fault if VIN goes below this (that is, we will not
+/// lose POWER_GOOD), but the part will indicate an input fault and pull
+/// PWR_CONT1_VCORE_TO_SP_ALERT_L low.
+///
+const VCORE_UV_WARN_LIMIT: units::Volts = units::Volts(11.75);
+
+///
+/// We want to collect enough samples (at ~900µs per sample) to adequately
+/// cover any anticipated dip.  We have seen these have an ~11ms total width
+/// in the wild, so we give ourselves plenty of margin here and get ~45ms
+/// of data.
+///
+const VCORE_NSAMPLES: usize = 50;
+
+const VCORE_TO_SP_ALERT_L: sys_api::PinSet = sys_api::Port::I.pin(14);
```

Comment:
Eventually, I'd like to wire up the `sys` task config for EXTI pins to some additional codegen so that client tasks can reference pins by their names in the config, instead of duplicating the definition. Not something we should change in this PR, just making a note of it.

Reply:
I've gone ahead and implemented this in https://github.com/oxidecomputer/hubris/pull/1734, so if that PR merges first, we can reuse this pin definition from the config.

Reply:
Okay, #1734 has merged, so (after a rebase) this can now get the pin set definition from `sys_api`: 
```suggestion
use sys_api::VCORE_TO_SP_ALERT_L;
```